app/App/Account/Account/Requests/TransactionRequest/TokenSpend/index.js

```diff
@@ -0,0 +1,262 @@
+import React from 'react'
+import Restore from 'react-restore'
+import BigNumber from 'bignumber.js'
+
+import svg from '../../../../../../../resources/svg'
+import link from '../../../../../../../resources/link'
+import { ClusterBox, Cluster, ClusterRow, ClusterValue } from '../../../../../../../resources/Components/Cluster'
+
+import { MAX_HEX } from '../../../../../../../resources/constants'
+import { formatDisplayInteger, isUnlimited } from '../../../../../../../resources/utils/numbers'
+
+class TokenSpend extends React.Component {
+  constructor (...args) {
+    super(...args)
+    this.state = {
+      inPreview: false,
+      inEditApproval: false,
+      mode: 'requested',
+      customInput: ''
+    }
+  }
+
+  setAmount (amount) {
+    this.setState({ amount })
+  }
+
+  setCustomAmount (value, decimals) {
+    if (value === '') {
+      this.setState({ mode: 'custom', amount: '0x0', customInput: value })
+    } else {
+      // only allow ints
+      if (!/^\d+$/.test(value)) return
+
+      const max = new BigNumber(MAX_HEX)
+      const custom = new BigNumber(value).shiftedBy(decimals)
+
+      let amount
+      if (max.comparedTo(custom) === -1) {
+        amount = MAX_HEX
+      } else {
+        amount = '0x' + custom.integerValue().toString(16)
+      }
+
+      this.setState({ mode: 'custom', amount, customInput: value })
+    }
+  }
+
+  startEditing () {
+    this.setState({ inEditApproval: true })
+  }
+
+  copySpenderAddress (data) {
+    link.send('tray:clipboardData', data)
+    this.setState({ copiedSpenderAddress: true })
+    setTimeout(() => this.setState({ copiedSpenderAddress: false }), 1000)
+  }
+
+  copyTokenAddress (data) {
+    link.send('tray:clipboardData', data)
+    this.setState({ copiedTokenAddress: true })
+    setTimeout(() => this.setState({ copiedTokenAddress: false }), 1000)
+  }
+
+  render () {
+    const {approval, updateApproval, requestedAmountHex } = this.props
+
+    const { data } = approval
+    const decimals = data.decimals || 0
+    const requestedAmount = requestedAmountHex
+    const customInput = '0x' + new BigNumber(this.state.customInput).shiftedBy(decimals).integerValue().toString(16)
+    const value = new BigNumber(data.amount)
+    const revoke = value.eq(0)
+
+    const displayAmount = isUnlimited(data.amount) ? 'unlimited' : formatDisplayInteger(data.amount, decimals)
+
+    const symbol = data.symbol || '???'
+    const name = data.name || 'Unknown Token'
+
+    const inputLock = !data.symbol || !data.name || !decimals
+
+    const spenderEns = data.spenderEns
+    const spender = data.spender
+
+    const tokenAddress = data.contract
```

Comment:
I think you could condense most of these data shortcut constants above like

```
const { name, amount, spenderEns, spender, decimals = 0, symbol = '???', name = 'Unknown Token' } = data
```

inputLock definitition would need to change.